src/json/lexer/io_based.cr

```diff
@@ -2,17 +2,64 @@
 class JSON::Lexer::IOBased < JSON::Lexer
   def initialize(@io : IO)
     super()
-    @current_char = @io.read_char || '\0'
+    @current_char = @io.read_byte.try(&.chr) || '\0'
   end
 
   private getter current_char
 
   private def next_char_no_column_increment
-    @current_char = @io.read_char || '\0'
+    @current_char = @io.read_byte.try(&.chr) || '\0'
   end
 
   private def consume_string
-    consume_string_with_buffer
+    peek = @io.peek
+    if !peek || peek.empty?
+      return consume_string_with_buffer
+    end
+
+    pos = 0
+
+    while true
```

Comment:
**thought:** I'm wondering if the strategy here could be based on byte search (`peek.index('"')`) instead? The implementation for that can be more efficient than iterating over each byte. `Slice#index` on a byte buffer is backed by `memchr`, so it depends on how much the libc implementation is optimized.

We still need to sanity check for escape sequences and unallowed characters in the potential string, though. So that would reduce effectiveness. I think it could overall be better that way, but I'm not sure. Just wanted to leave this thought here. It should be good to take the current implementation for now.

Reply:
Might be true. When I needed to find the end of a JSON string in a `Bytes`, I came up with something like this:

```crystal
def json_bytes_end_of_string_index( haystack : Bytes, offset : Int32 = 0 ) : Int32?
  index = haystack.index( '"'.ord.to_u8, offset )  # memchr()
  return nil  if ! index
  return index  unless haystack[ index - 1 ] == '\\'.ord.to_u8
  # Fall back to the more complete bytewise implementation below.
  # ...
end
```

I did not check if it makes sense to use `strpbrk`() to find the next "interesting" byte, such as quote, backslash, ...

Reply:
That's the other optimization I was thinking. It makes string parsing much faster. I didn't want to include it in this PR though!